### PR TITLE
Alit/mamba

### PR DESCRIPTION
**Pull request by JRD971000**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/text_generation_strategy.py [hunk 1]
```diff
@@ -981,6 +1057,7 @@
         MegatronGPTPromptLearningModel,
     )
     from nemo.collections.nlp.models.language_modeling.megatron_griffin_model import MegatronGriffinModel
+    from nemo.collections.nlp.models.language_modeling.megatron_mamba_model import MegatronMambaModel
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply nemo/collections/nlp/modules/common/text_generation_strategy.py:984]: can you check this cycle issue?

File: nemo/collections/nlp/parts/mixins/nlp_adapter_mixins.py [hunk 2]
```diff
@@ -178,9 +179,10 @@ def _check_and_add_peft_cfg(self, peft_cfg):
                 for layer in layers:
                     if layer.layer_number in (layer_selection or list(range(1, self.cfg.num_layers + 1))):
                         for name, module in layer.named_modules():
-                            self._check_and_add_adapter(
-                                name, module, adapter_name, adapter_cfg, name_key_to_mcore_mixins
-                            )
+                            if not isinstance(module, IdentityOp):
+                                self._check_and_add_adapter(
```
Comment (yaoyu-33) [inline nemo/collections/nlp/parts/mixins/nlp_adapter_mixins.py:183]: is it better to add this check inside `self._check_and_add_adapter`?

File: nemo/collections/nlp/models/language_modeling/megatron_mamba_sft_model.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+from omegaconf import DictConfig
+from omegaconf.dictconfig import DictConfig
+from pytorch_lightning.trainer.trainer import Trainer
+
+from nemo.collections.nlp.models.language_modeling.megatron_base_model import MegatronBaseModel
+from nemo.collections.nlp.models.language_modeling.megatron_gpt_sft_model import MegatronGPTSFTModel
+from nemo.collections.nlp.models.language_modeling.megatron_mamba_model import MegatronMambaModel
+
+try:
+    HAVE_APEX = True
```
Comment (yaoyu-33) [inline nemo/collections/nlp/models/language_modeling/megatron_mamba_sft_model.py:24]: not importing anything here, remove

File: nemo/collections/nlp/models/language_modeling/megatron_mamba_sft_model.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        if not HAVE_APEX:
+            raise ImportError(
+                "Apex was not found. Please see the NeMo README for installation instructions: https://github.com/NVIDIA/NeMo#megatron-gpt."
+            )
+
+        super().__init__(cfg, trainer=trainer)
+        self.mcore_gpt = True
```
Comment (yaoyu-33) [inline nemo/collections/nlp/models/language_modeling/megatron_mamba_sft_model.py:43]: why? this should from cfg right?

Comment (yaoyu-33) [reply nemo/collections/nlp/models/language_modeling/megatron_mamba_sft_model.py:43]: if you are overriding at least throw a warning for this

File: examples/nlp/language_modeling/mamba_change_num_partition.py [hunk 5]
```diff
@@ -0,0 +1,716 @@
...
+
+            param.data = split_val
+            idx += 1
+
+        if write_path is not None:
+            logging.info(f"Writing pp rank {pp_rank} tp rank {tp_rank} to file {write_path}")
+            model.save_to(write_path)
+
+
+def debug_log_split_param_diff(idx, param, param_name, partitions):
```
[the remark this answers is not in the document]
Comment (JRD971000) [reply examples/nlp/language_modeling/mamba_change_num_partition.py:262]: done

File: scripts/checkpoint_converters/convert_mamba2_pyt_to_nemo.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        index_str = str(i)
+        layer_types = layers.get(index_str, [])
+        if 'mixer' in layer_types:
+            layer_pattern += 'M'
+        elif 'self_attention' in layer_types:
+            layer_pattern += '*'
+        elif 'mlp' in layer_types:
+            layer_pattern += '-'
+        else:
+            AssertionError("Layer not found. Each layer must be eiher MLP, Mamba, or Attention")
```
[the remark this answers is not in the document]
Comment (JRD971000) [reply scripts/checkpoint_converters/convert_mamba2_pyt_to_nemo.py]: fixed

File: nemo/collections/nlp/parts/mixins/nlp_adapter_mixins.py (hunk 2 above)
Comment (JRD971000) [reply nemo/collections/nlp/parts/mixins/nlp_adapter_mixins.py:183]: Done

File: nemo/collections/nlp/models/language_modeling/megatron_mamba_sft_model.py (hunk 4 above)
Comment (JRD971000) [reply nemo/collections/nlp/models/language_modeling/megatron_mamba_sft_model.py:35]: Added a warning

File: nemo/collections/nlp/modules/common/text_generation_strategy.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JRD971000) [reply nemo/collections/nlp/modules/common/text_generation_strategy.py:991]: I think this import cycle is already there for GPT (and other models). For example, for GOT, it is imported in the text gen staratgy here: 

https://github.com/NVIDIA/NeMo/blob/d4a32d0dea3d7201defdad09967b4536fa56e672/nemo/collections/nlp/modules/common/text_generation_strategy.py#L986 

And MegatronGPTModel imports text gen strategy here:
https://github.com/NVIDIA/NeMo/blob/d4a32d0dea3d7201defdad09967b4536fa56e672/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L57

**Review by yaoyu-33 — APPROVED** [review]

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**